### Import/load data

In [76]:
import pandas as pd

In [77]:
df = pd.read_csv('WA_Fn-UseC_-Telco-Customer-Churn.csv')
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


### Check the data

In [78]:
df.shape

(7043, 21)

In [79]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


TotalCharges data type should be float instead of object

There is/are some data in column TotalCharges that has value ' '

In [80]:
df[df['TotalCharges'] == ' ']

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
488,4472-LVYGI,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,Yes,No,Two year,Yes,Bank transfer (automatic),52.55,,No
753,3115-CZMZD,Male,0,No,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.25,,No
936,5709-LVOEQ,Female,0,Yes,Yes,0,Yes,No,DSL,Yes,...,Yes,No,Yes,Yes,Two year,No,Mailed check,80.85,,No
1082,4367-NUYAO,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.75,,No
1340,1371-DWPAZ,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,Yes,No,Two year,No,Credit card (automatic),56.05,,No
3331,7644-OMVMY,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,19.85,,No
3826,3213-VVOLG,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.35,,No
4380,2520-SGTTA,Female,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.00,,No
5218,2923-ARZLG,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,One year,Yes,Mailed check,19.70,,No
6670,4075-WKNIU,Female,0,Yes,Yes,0,Yes,Yes,DSL,No,...,Yes,Yes,Yes,No,Two year,No,Mailed check,73.35,,No


TotalCharges is ' ' for 11 customers with tenure = 0 (newly registered, not yet billed). So, replace with 0 since no charges have accumulated yet

In [81]:
df['TotalCharges'] = df['TotalCharges'].replace(' ', '0')

In [82]:
df[df['TotalCharges'] == ' ']

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn


In [83]:
# Change column 'TotalCharges' data type to float
df['TotalCharges'] = df['TotalCharges'].astype(float)

In [84]:
df['TotalCharges'].dtypes

dtype('float64')

In [85]:
df.describe()

,SeniorCitizen,tenure,MonthlyCharges,TotalCharges
count,7043.000000,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692,2279.734304
std,0.368612,24.559481,30.090047,2266.794470
min,0.000000,0.000000,18.250000,0.000000
25%,0.000000,9.000000,35.500000,398.550000
50%,0.000000,29.000000,70.350000,1394.550000
75%,0.000000,55.000000,89.850000,3786.600000
max,1.000000,72.000000,118.750000,8684.800000


In [86]:
df['Churn'].value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

### Train test split

In [87]:
df.head(3)

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.50,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes


In [88]:
# Seperate X and y
X = df.drop(columns=['customerID', 'Churn'])
y = df['Churn']

In [89]:
X.head(3)

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges
0,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85
1,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.50
2,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15


In [90]:
y.head(3)

0     No
1     No
2    Yes
Name: Churn, dtype: object

In [91]:
# train test split
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=88)

In [92]:
X_train.shape

(5634, 19)

In [93]:
X_test.shape

(1409, 19)

In [94]:
y_train.shape

(5634,)

In [95]:
y_test.shape

(1409,)

### Data preprocessing and building pipeline

In [96]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, StandardScaler

In [97]:
X_train.head(3)

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges
875,Female,0,No,No,3,No,No phone service,DSL,No,No,No,No,Yes,No,Month-to-month,Yes,Credit card (automatic),34.80,113.95
1025,Female,1,Yes,No,60,Yes,No,Fiber optic,Yes,Yes,Yes,Yes,Yes,Yes,Two year,Yes,Bank transfer (automatic),110.00,6668.35
947,Female,1,Yes,Yes,32,Yes,Yes,Fiber optic,No,No,Yes,No,Yes,No,Month-to-month,Yes,Mailed check,91.35,2896.55


In [98]:
# Preprocessing
binary_cols  = ['gender', 'Partner', 'Dependents', 'PhoneService', 'PaperlessBilling']
ordinal_cols = ['Contract']
nominal_cols = ['MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup',
                'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies',
                'PaymentMethod']
numeric_cols = ['tenure', 'MonthlyCharges', 'TotalCharges', 'SeniorCitizen']

preprocessor = ColumnTransformer([
    ('bin', OrdinalEncoder(), binary_cols),
    ('ord', OrdinalEncoder(categories=[['Month-to-month', 'One year', 'Two year']]), ordinal_cols),
    ('nom', OneHotEncoder(drop='first', handle_unknown='ignore'), nominal_cols),
    ('num', StandardScaler(), numeric_cols),
])


In [99]:
# Pipeline
from sklearn.ensemble import RandomForestClassifier

model = Pipeline([
    ('preprocessor', preprocessor),
    ('classifier', RandomForestClassifier(max_depth=8, random_state=88))
]) # manually chosen hyperparameter

### Training and evaluation

In [100]:
model.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('bin', OrdinalEncoder(),
                                                  ['gender', 'Partner',
                                                   'Dependents', 'PhoneService',
                                                   'PaperlessBilling']),
                                                 ('ord',
                                                  OrdinalEncoder(categories=[['Month-to-month',
                                                                              'One '
                                                                              'year',
                                                                              'Two '
                                                                              'year']]),
                                                  ['Contract']),
                                                 ('nom',
                                                  OneHotEncoder(drop='first',
                                                                handle_unknown='ignore'),
                                                  ['MultipleLines',
                                                   'InternetService',
                                                   'OnlineSecurity',
                                                   'OnlineBackup',
                                                   'DeviceProtection',
                                                   'TechSupport', 'StreamingTV',
                                                   'StreamingMovies',
                                                   'PaymentMethod']),
                                                 ('num', StandardScaler(),
                                                  ['tenure', 'MonthlyCharges',
                                                   'TotalCharges',
                                                   'SeniorCitizen'])])),
                ('classifier',
                 RandomForestClassifier(max_depth=8, random_state=88))])

In [101]:
y_pred = model.predict(X_test)

In [102]:
# Evaluation

from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred, digits=4))

              precision    recall  f1-score   support

          No     0.8341    0.9139    0.8722      1034
         Yes     0.6775    0.4987    0.5745       375

    accuracy                         0.8034      1409
   macro avg     0.7558    0.7063    0.7233      1409
weighted avg     0.7924    0.8034    0.7929      1409



### Comparing Random Forest performance with Decision Tree performance

In [103]:
# Pipeline
from sklearn.tree import DecisionTreeClassifier

model2 = Pipeline([
    ('preprocessor', preprocessor),
    ('classifier', DecisionTreeClassifier(max_depth=6, random_state=88))
]) # manually chosen hyperparameter

In [104]:
model2.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('bin', OrdinalEncoder(),
                                                  ['gender', 'Partner',
                                                   'Dependents', 'PhoneService',
                                                   'PaperlessBilling']),
                                                 ('ord',
                                                  OrdinalEncoder(categories=[['Month-to-month',
                                                                              'One '
                                                                              'year',
                                                                              'Two '
                                                                              'year']]),
                                                  ['Contract']),
                                                 ('nom',
                                                  OneHotEncoder(drop='first',
                                                                handle_unknown='ignore'),
                                                  ['MultipleLines',
                                                   'InternetService',
                                                   'OnlineSecurity',
                                                   'OnlineBackup',
                                                   'DeviceProtection',
                                                   'TechSupport', 'StreamingTV',
                                                   'StreamingMovies',
                                                   'PaymentMethod']),
                                                 ('num', StandardScaler(),
                                                  ['tenure', 'MonthlyCharges',
                                                   'TotalCharges',
                                                   'SeniorCitizen'])])),
                ('classifier',
                 DecisionTreeClassifier(max_depth=6, random_state=88))])

In [105]:
y_pred2 = model2.predict(X_test)

In [106]:
# Decision Tree
print(classification_report(y_test, y_pred2, digits=4))

              precision    recall  f1-score   support

          No     0.8307    0.9014    0.8646      1034
         Yes     0.6446    0.4933    0.5589       375

    accuracy                         0.7928      1409
   macro avg     0.7376    0.6973    0.7117      1409
weighted avg     0.7811    0.7928    0.7832      1409



In [107]:
# Random Forest
print(classification_report(y_test, y_pred, digits=4))

              precision    recall  f1-score   support

          No     0.8341    0.9139    0.8722      1034
         Yes     0.6775    0.4987    0.5745       375

    accuracy                         0.8034      1409
   macro avg     0.7558    0.7063    0.7233      1409
weighted avg     0.7924    0.8034    0.7929      1409



### Hyperparameter Tuning with Grid Search and Stratified Cross-Validation

In [108]:
from sklearn.model_selection import GridSearchCV, StratifiedKFold
from sklearn.metrics import make_scorer, f1_score

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=88)
f1_churn = make_scorer(f1_score, pos_label='Yes')

rf_search = GridSearchCV(
    model,
    param_grid={
        'classifier__n_estimators': [100, 300],
        'classifier__max_depth': [4, 6, 8, 10, 12, None],
        'classifier__min_samples_leaf': [1, 5, 10],
    },
    scoring=f1_churn, cv=cv, n_jobs=-1
)
rf_search.fit(X_train, y_train)

dt_search = GridSearchCV(
    model2,
    param_grid={
        'classifier__max_depth': [3, 4, 5, 6, 8, 10],
        'classifier__min_samples_leaf': [1, 5, 10, 20],
    },
    scoring=f1_churn, cv=cv, n_jobs=-1
)
dt_search.fit(X_train, y_train)

print(rf_search.best_params_, rf_search.best_score_)
print(dt_search.best_params_, dt_search.best_score_)

print(' ')
print('Random Forest')
print(classification_report(y_test, rf_search.predict(X_test), digits=4))

print(' ')
print('Decision Tree')
print(classification_report(y_test, dt_search.predict(X_test), digits=4))

{'classifier__max_depth': 12, 'classifier__min_samples_leaf': 5, 'classifier__n_estimators': 100} 0.5817796603473657
{'classifier__max_depth': 5, 'classifier__min_samples_leaf': 1} 0.5762693862238522
 
Random Forest
              precision    recall  f1-score   support

          No     0.8339    0.9081    0.8694      1034
         Yes     0.6643    0.5013    0.5714       375

    accuracy                         0.7999      1409
   macro avg     0.7491    0.7047    0.7204      1409
weighted avg     0.7888    0.7999    0.7901      1409

 
Decision Tree
              precision    recall  f1-score   support

          No     0.8224    0.9004    0.8596      1034
         Yes     0.6282    0.4640    0.5337       375

    accuracy                         0.7842      1409
   macro avg     0.7253    0.6822    0.6967      1409
weighted avg     0.7707    0.7842    0.7729      1409



In [109]:
y_test.value_counts(normalize=True)

Churn
No     0.733854
Yes    0.266146
Name: proportion, dtype: float64

Conclusion:
- After hyperparameter tuning, Random Forest performed better than Decision Tree on the test set in all reported metrics
- Both models are good at identifying customers who stay (recall around 0.90), but still miss about half of the customers who churn (recall around 0.46 to 0.50)
- Accuracy of around 0.80 is only slightly above the majority-class baseline (0.73), so churn recall and F1-score are more meaningful metrics for this imbalanced data.